### PR TITLE
internal/extsvc/github: Cache oauth scopes in redis with 5s TTL

### PR DESCRIPTION
**Pull request by indradhanush**
In POC stage. Looking for some early feedback.

Maybe helps reduce the high volume of GitHub API requests observed in https://github.com/sourcegraph/customer/issues/658.

### CODE REVIEW

File: internal/extsvc/github/v3.go [hunk 1]
```diff
@@ -406,16 +418,27 @@ func (c *V3Client) GetAuthenticatedOAuthScopes(ctx context.Context) ([]string, e
...
 		return nil, err
 	}
+
 	scope := header.Get("x-oauth-scopes")
 	if scope == "" {
 		return []string{}, nil
 	}
+
+	// Set the scope in the cache.
+	c.oauthScopesCache.Set(key, []byte(scope))
```
Comment (indradhanush) [inline internal/extsvc/github/v3.go:440]: [Open question](https://sourcegraph.slack.com/archives/C07KZF47K/p1643896131884479) right now I have is whether this is safe for concurrent access.

**Comment by indradhanush** [comment]
@bobheadxi I read more of the code around how `Validate` is being invoked at the moment and here are my thoughts and findings. 

From [enterprise/cmd/worker/main.go](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7c691ff15853058c5e8b981d10cf8d0dd5fd240a/-/blob/enterprise/cmd/worker/main.go?L53-57&subtree=true#tab=references), we have this helpful comment which perhaps explains why we need to keep refreshing the authz providers (and unintentionally end up making the API calls to GitHub):

>// setAuthProviders waits for the database to be initialized, then periodically refreshes the
// global authz providers. This changes the repositories that are visible for reads based on the
// current actor stored in an operation's context, which is likely an internal actor for many of
// the jobs configured in this service. This also enables repository update operations to fetch
// permissions from code hosts.

Arguably we could increase this interval (and in other places), but I'm not sure what side effects this could potentially lead to. I'm hoping @efritz can provide more insight. :thinking: 

Admittedly, I do not yet understand how refreshing the authz provider every 5 seconds affects the repo visibility for each actor? :thinking: 

Coming back to `Validate`, here's how the call stack goes:

[ProvidersFromConfig](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7c691ff15853058c5e8b981d10cf8d0dd5fd240a/-/blob/enterprise/internal/authz/authz.go?L32:6&subtree=true#tab=references) -> [github.NewAuthzProviders](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7c691ff15853058c5e8b981d10cf8d0dd5fd240a/-/blob/enterprise/internal/authz/github/authz.go?L20:6&subtree=true#tab=references) -> [authz.Provider.Validate](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7c691ff15853058c5e8b981d10cf8d0dd5fd240a/-/blob/enterprise/internal/authz/github/authz.go?L79&subtree=true)

As a result I don't see an easy way to implement the `ValidateConnection` approach suggested above that could immediately benefit the callers of `ProvidersFromConfig`. Although I think that it would have been the right solution for the long term but would require a larger scale refactor and potentially also has a greater area of impact.

Important to note that github.NewAuthzProviders is also invoked from [ProviderFromExternalService](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7c691ff15853058c5e8b981d10cf8d0dd5fd240a/-/blob/enterprise/internal/authz/authz.go?L183:6&subtree=true#tab=references) thus also being a source of indirectly invoking `authz.Provider.Validate` and thus making API calls to GitHub. However, the good news is that `ProviderFromExternalService` is only invoked from the perms syncer which keeps the origin of these requests within `repo-updater`. But with the caching changes in this PR these API calls could also benefit from them.

> Im not sure this will solve the issue of periodic instances of `Validate()`ing loads of authz configs, which seems to be the issue here judging from the Slack threads - in these situations, every call to get the oauth scopes will be a cache miss unless we make the TTL multiple hours.

Rephrasing our discussion on Slack to preserve context, unless each service is making the API requests in the same second with their internal clocks synchronized I'm hopeful that we see a high ratio of cache hits to misses. The key used is a hash of the authenticator embedded in the provider's client and unless the underlying API token in use is different, all the services should be able to get cache hits more often than not. 

I'm inclined to give this approach a try and see if it helps alleviate the stress on the GitHub API with respect to number of API calls we're making across all our services. I've added prometheus metrics to track this so it should be pretty straightforward to find out the results.

What do you think? I'm open to feedback and discussions. :smile:

**Comment by efritz** [comment]
> Arguably we could increase this interval (and in other places), but I'm not sure what side effects this could potentially lead to. I'm hoping @efritz can provide more insight. Admittedly, I do not yet understand how refreshing the authz provider every 5 seconds affects the repo visibility for each actor?

As I understand it (from a high-level, so I could be wrong), we are updating _and caching_ the oauth scopes available to the configured extsvc. This means that every five seconds this cache gets updated, so any operations within that window will see the previous refresh.

The precise-code-intel-worker is always operating at system-level, though. I'm sure we could up this interval in this service (specifically). I'd opt to even make it configurable for all services with some sane default and a doc page about increasing/decreasing the interval.

I'm not super familiar with the authz mechanisms, though. I'd refer you to @unknwon, as he was the one originally doing the work to support this. I think the effect of _removing_ it makes everything invisible to the precise code intelligence upload processor, which is why it was added in the first place.

**Comment by indradhanush** [comment]
> As I understand it (from a high-level, so I could be wrong), we are updating and caching the oauth scopes available to the configured extsvc. This means that every five seconds this cache gets updated, so any operations within that window will see the previous refresh.

@efritz That is also similar to my understanding. What I have failed to figure out yet is though why do we need to cache the oauth scope? If a user removes a scope from their GitHub dashboard, subsequent queries from Sourcegraph would fail immediately if that scope was required. I was chatting with @bobheadxi yesterday about this and couldn't figure out why do we need to update the oauth scope in our cache in all the other services apart from `frontend`, which seems like the only place we need this behaviour to let the user know in case their token is missing required scopes.

**Comment by efritz** [comment]
@indradhanush 

> If a user removes a scope from their GitHub dashboard, subsequent queries from Sourcegraph would fail immediately if that scope was required

Why? Wouldn't we still serve data that should be now non-visible to that user from Sourcegraph? Isn't the point of the cache **not** to check with the authority on every request?

**Comment by indradhanush** [comment]
> Why? Wouldn't we still serve data that should be now non-visible to that user from Sourcegraph? Isn't the point of the cache not to check with the authority on every request?

@efritz Ah of course! Now it makes sense. I shoe-horned myself to only look at this from the perspective of syncing data to Sourcegraph from GitHub and never considered the part of serving the data our to Sourcegraph's users. I knew I was missing something obvious. This helps me tremendously. Thank you so much! :pray:

**Comment by indradhanush** [comment]
@efritz So the other question is if we also need to be as pedantic as to update the scope every 5 seconds? Does making it 5s by default and letting the site admin configure this sound reasonable? (That would be a separate PR of course)

**Comment by efritz** [comment]
@indradhanush Yes! <3 That's what I suggested above:

> The precise-code-intel-worker is always operating at system-level, though. I'm sure we could up this interval in this service (specifically). I'd opt to even make it configurable for all services with some sane default and a doc page about increasing/decreasing the interval.

I'd definitely run this by someone with more confidence about authz requirements, though. I'm just an observer 🤓

**Comment by bobheadxi** [comment]
> As a result I don't see an easy way to implement the ValidateConnection approach suggested above that could immediately benefit the callers of ProvidersFromConfig. Although I think that it would have been the right solution for the long term but would require a larger scale refactor and potentially also has a greater area of impact.

`ProvidersFromConfig` could simply introduce a parameter that toggles whether `ValidateConnection` is done. Most services can get away with just a simple `Validate`, and the frontend can `ValidateConnection` in response to certain requests. Or, the refresh job could enable this parameter once every x seconds

**Comment by indradhanush** [comment]
> ProvidersFromConfig could simply introduce a parameter that toggles whether ValidateConnection is done. Most services can get away with just a simple Validate, and the frontend can ValidateConnection in response to certain requests. Or, the refresh job could enable this parameter once every x seconds

I think it makes the API of `ProvidersFromConfig` clunky by introducing a toggle for what appears to be an implementation level detail. We should rather have two separate top level APIs so that it's always clear to the consumer of the method which one they're using and why, instead of having to read the code first. What do you think?

**Comment by unknwon** [comment]
Seeing several topics here, trying my best:

1. @bobheadxi IIRC Is the historical of having OAuth scopes checked solely for better debugging purpose (for group membership caching)? Correct me if I'm wrong but if yes, I'm curious to know whether we could get this checking only on run the on-demand (e.g. only run on frontend) vs within the implementation of the `Validate` method. 
	- The suggestion of `ValidateConnection` is one way of doing this, but this is changing a very abstract interface, and the call of `Validate` (as @indradhanush posted) buried deep under a call chain, and with our current code structure, I'm not sure if it's even possible to pass `true/false` at the right time without non-small refactoring. Personally I'd love it to be a method specifically about GitHub, just my 2 cents though.
1. @indradhanush The interval of calls of `ProvidersFromConfig` determines the responsiveness of how fast our system reacts to the changes of authz provider in any external services, ideally it should be 0 second and I take 5 seconds as a good compromise due to limitation of current authz architecture (pulling).
1. @indradhanush I start to wonder how much value we get from checking OAuth scopes once the OAuth access token has been granted to us because it rarely changes (customer admin misconfigured). Correlates to my bullet 1 above, I think checking is more of debugging and making TTL to 1 hour would still make sense to me.

File: internal/extsvc/github/v3.go [hunk 2]
```diff
@@ -222,6 +226,15 @@ func newRepoCache(apiURL *url.URL, a auth.Authenticator) *rcache.Cache {
 	return rcache.NewWithTTL("gh_repo:"+key, int(cacheTTL/time.Second))
 }
 
+func newOAuthScopesCache(a auth.Authenticator) *rcache.Cache {
+	key := ""
```
Comment (unknwon) [inline internal/extsvc/github/v3.go:230]: Do we wanna having key like `gh_oauth_scopes:` for no authenticator or just skip the cache entirely?

Comment (indradhanush) [reply internal/extsvc/github/v3.go:230]: That's a good catch. I think we should skip the cache. Will fix.

**Comment by bobheadxi** [comment]
> Correct me if I'm wrong but if yes, I'm curious to know whether we could get this checking only on run the on-demand (e.g. only run on frontend) vs within the implementation of the Validate method.

Agreed - and I agree the passing-argument-down is probably too much refactoring. Overall I believe the `ValidateConnection` approach is the best way to split out this logic.

> [...] but this is changing a very abstract interface [...] Personally I'd love it to be a method specifically about GitHub, just my 2 cents though.

Not a huge fan of this since it leaks the abstraction, and I think `ValidateConnection` does have a good fit in the interface: it fits into the Provider abstraction as a connection check and not a config check, and could enable us to present auth errors better up-front if tied into the right place. It also side-steps the issue of:

> the call of Validate [...] buried deep under a call chain

because it would not require a significant refactor - all calls to `Validate` can stay the same, except now they will be free of network connections again, and we just have to pinpoint the single place in the code where the frontend presents issues to the users and add a new call there.

**Comment by mollylogue** [comment]
I'm taking this over from Indra as I'm the support person for repo-management this week. I'm planning to open a separate PR w/ configurable refresh period for authz. I'm a bit confused on some of the comments here and just trying to understand what the best path forward is. I think it's worthwhile to try the cache to see what kind of difference it makes. I don't think I totally understand your recommendation @bobheadxi about adding a second function `ValidateConnection`. It looks to me like making an API call to github is a crucial part of what `Validate` does. Maybe I'm misunderstanding your recommendation.

**Comment by bobheadxi** [comment]
> It looks to me like making an API call to github is a crucial part of what Validate does

The API call ensures the appropriate scopes have been requested - if not provided auth issues will ensue, but it is not critical for every services to check all the time, since services that reach out to code hosts should be resilient to auth issues. However, it does provide an avenue through which we can provide useful feedback to the user about what might be missing.

My thought is that splitting the heavier validation can happen only when we're trying to show something to the user and not everywhere - hence a new interface for conducting this type of validation, `ValidateConnection(context.Context)`

**Comment by mollylogue** [comment]
> > It looks to me like making an API call to github is a crucial part of what Validate does
> 
> The API call ensures the appropriate scopes have been requested - if not provided auth issues will ensue, but it is not critical for every services to check all the time, since services that reach out to code hosts should be resilient to auth issues. However, it does provide an avenue through which we can provide useful feedback to the user about what might be missing.
> 
> My thought is that splitting the heavier validation can happen only when we're trying to show something to the user and not everywhere - hence a new interface for conducting this type of validation, `ValidateConnection(context.Context)`

Okay. I guess I'm still not understanding what other validation `Validate()` is currently doing. It seems that you're advocating for just not calling `Validate()` in the case that we aren't calling from the front end? It looks to me, based on reading through what Validate does [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/authz/github/github.go?L91:20) it's comparing the required auth scopes with what we get from github. I'm just really confused as to what this would be validating without the `scopes` returned from github.

**Comment by bobheadxi** [comment]
> Okay. I guess I'm still not understanding what other validation Validate() is currently doing. It seems that you're advocating for just not calling Validate() in the case that we aren't calling from the front end?

No, just that we shouldn't do this specific problematic check (network request for scopes)

> It looks to me, based on reading through what Validate does [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/authz/github/github.go?L91:20) it's comparing the required auth scopes with what we get from github. I'm just really confused as to what this would be validating without the scopes returned from github.

That's just GitHub - [there are many implementations of Validate() that do more extensive checks](https://sourcegraph.com/search?q=context:%40sourcegraph/all+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+file:%5Eenterprise/internal/authz+func+%28...%29+Validate%28%29+...+%7B+...+%7D+count:all&patternType=structural) to satisfy [the Provider interface](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/authz/iface.go#L83:6-83:14).

However, looking at this I realized that [every single non-no-op implementation of `Validate()` makes network requests](https://sourcegraph.com/search?q=context:%40sourcegraph/all+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+file:%5Eenterprise/internal/authz+func+%28...%29+Validate%28%29+...+%7B+...+%7D+count:all&patternType=structural):

- Perforce: p4 exec
- Bitbucket server: gets client username and perms
- Gitlab sudo: gets token

There is _no implementation of `Validate` that does the "quick check"_ I imagined [might be the usual case](https://github.com/sourcegraph/sourcegraph/pull/30617#issuecomment-1030683151) 😅  I think we need to completely change the way `Valdiate()` gets used such that it is not in the path of so many things, and allow it to accept `context.Context`

**Comment by mollylogue** [comment]
@bobheadxi What do you think about trying this caching approach for now as a short-term solution (and validating that it actually makes a difference i.e. that we're hitting the cache and reducing the number of api requests) and saving the larger refactor for the future? I agree that what you're describing is the correct approach but I also think we need a quick win for the customer if possible.

**Comment by bobheadxi** [comment]
@mollylogue this is a quick pull request demonstrates what I think is the right solution: https://github.com/sourcegraph/sourcegraph/pull/30847 - I've included some searches indicating that all the calls today to `Validate` are superfluous. Happy to sync over this approach!

**Comment by mollylogue** [comment]
FYI y'all I think we're going to go with the approach outlined in [this PR](https://github.com/sourcegraph/sourcegraph/pull/30847#pullrequestreview-876552325) in which case we won't need this extra bit of caching. I'm going to leave this open until we resolve the issue with the customer (just in case) and then I'll just close this PR.

**Pull request closed without merging**